pkg/providers/securitygroup/securitygroup.go

```diff
@@ -0,0 +1,150 @@
+}
+
+func NewDefaultProvider(region string, ecsapi ecs.Client, cache *cache.Cache) *DefaultProvider {
+	return &DefaultProvider{
+		region: region,
+		ecsapi: ecsapi,
+		cm:     pretty.NewChangeMonitor(),
+		// TODO: Remove cache cache when we utilize the security groups from the ECSNodeClass.status
+		cache: cache,
+	}
+}
+
+func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1alpha1.ECSNodeClass) ([]*ecs.DescribeSecurityGroupsResponseBodySecurityGroupsSecurityGroup, error) {
+	p.Lock()
+	defer p.Unlock()
+
+	// Get SecurityGroups
+	filterSets := getFilterSets(nodeClass.Spec.SecurityGroupSelectorTerms)
+	securityGroups, err := p.getSecurityGroups(filterSets)
+	if err != nil {
+		return nil, err
+	}
+	securityGroupIDs := lo.Map(securityGroups, func(s *ecs.DescribeSecurityGroupsResponseBodySecurityGroupsSecurityGroup, _ int) string {
+		return *s.SecurityGroupId
+	})
+	if p.cm.HasChanged(fmt.Sprintf("security-groups/%s", nodeClass.Name), securityGroupIDs) {
+		log.FromContext(ctx).
+			WithValues("security-groups", securityGroupIDs).
+			V(1).Info("discovered security groups")
+	}
+	return securityGroups, nil
+}
+
+func (p *DefaultProvider) getSecurityGroups(filterSets []*ecs.DescribeSecurityGroupsRequest) ([]*ecs.DescribeSecurityGroupsResponseBodySecurityGroupsSecurityGroup, error) {
+	hash, err := hashstructure.Hash(filterSets, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
+	if err != nil {
+		return nil, err
+	}
+	if sg, ok := p.cache.Get(fmt.Sprint(hash)); ok {
+		// Ensure what's returned from this function is a shallow-copy of the slice (not a deep-copy of the data itself)
+		// so that modifications to the ordering of the data don't affect the original
+		return append([]*ecs.DescribeSecurityGroupsResponseBodySecurityGroupsSecurityGroup{}, sg.([]*ecs.DescribeSecurityGroupsResponseBodySecurityGroupsSecurityGroup)...), nil
+	}
+	securityGroups := map[string]*ecs.DescribeSecurityGroupsResponseBodySecurityGroupsSecurityGroup{}
+	for _, filter := range filterSets {
+		if err := p.describeSecurityGroups(filter, func(securityGroup *ecs.DescribeSecurityGroupsResponseBodySecurityGroupsSecurityGroup) {
+			securityGroups[lo.FromPtr(securityGroup.SecurityGroupId)] = securityGroup
+		}); err != nil {
+			return nil, fmt.Errorf("describing security groups %+v, %w", filter, err)
+		}
+	}
+	p.cache.SetDefault(fmt.Sprint(hash), lo.Values(securityGroups))
+	return lo.Values(securityGroups), nil
+}
+
+func (p *DefaultProvider) describeSecurityGroups(request *ecs.DescribeSecurityGroupsRequest, process func(*ecs.DescribeSecurityGroupsResponseBodySecurityGroupsSecurityGroup)) error {
+	runtime := &util.RuntimeOptions{}
+	request.RegionId = tea.String(p.region)
+	request.MaxResults = tea.Int32(100)
+	request.IsQueryEcsCount = tea.Bool(true)
+	for {
+		output, err := p.ecsapi.DescribeSecurityGroupsWithOptions(request, runtime)
+		if err != nil {
+			return err
```

Comment:
log out the error here? The same goes for anything else that has an error

Reply:
The error is thrown to the outer handler, where it is printed. So I don't think it's necessary to print it separately.